lte/gateway/c/core/oai/tasks/nas/emm/EmmCommon.c

```diff
@@ -539,3 +539,208 @@ partial_list_t* emm_verify_orig_tai(const tai_t orig_tai) {
+            par_tai_list->tac[itr];
+      }
+      break;
+    default:
+      OAILOG_ERROR_UE(
+          imsi64, LOG_NAS,
+          "BAD TAI list configuration, unknown TAI list type %u",
+          par_tai_list->list_type);
+  }
+
+  /* TS 124.301 V15.4.0 Section 9.9.3.33:
+   * "The Tracking area identity list is a type 4 information element,
+   * with a minimum length of 8 octets and a maximum length of 98 octets.
+   * The list can contain a maximum of 16 different tracking area identities."
+   * We will limit the number to 1 partial list which can have maximum of 16
+   * TAIs.
+   */
+  tai_list->numberoflists = 1;
+  OAILOG_INFO_UE(
+      LOG_NAS, imsi64,
+      "  Got GUTI " GUTI_FMT ". The number of TAI partial lists: %d",
+      GUTI_ARG(&guti), tai_list->numberoflists);
+  OAILOG_FUNC_RETURN(LOG_NAS_EMM, RETURNok);
+}
+
+/****************************************************************************
+ **                                                                        **
+ ** Name:        verify_tau_tai                                            **
+ **                                                                        **
+ ** Description: Verifies if the TAI received during TAU                   **
+ **              is configured                                             **
+ **                                                                        **
+ ** Inputs:      imsi,guti,                                                **
+ **              tai: TAI received in TAU                                  **
+ **              emm_ctx_tai: pointer to tai_list_t in emm context         **
+ **                                                                        **
+ ** Outputs:     None                                                      **
+ **      Return:    RETURNok, RETURNerror                                  **
+ **      Others:    None                                                   **
+ **                                                                        **
+ ***************************************************************************/
+status_code_e verify_tau_tai(
+    uint64_t imsi64, guti_t guti, tai_t tai, tai_list_t* emm_ctx_tai) {
+  OAILOG_FUNC_IN(LOG_NAS_EMM);
+  /* Check if the TAI matches with the TAI in emm context.
+   * If it does not match, check if it matches with one of the partial
+   * TAI lists stored in mme_config and update the new TAI to emm context.
+   * Note that there is only one partial list stored in emm context.
+   */
+  switch (emm_ctx_tai->partial_tai_list[0].typeoflist) {
+    case TRACKING_AREA_IDENTITY_LIST_TYPE_ONE_PLMN_CONSECUTIVE_TACS:
+      if ((IS_PLMN_EQUAL(
+              tai.plmn, emm_ctx_tai->partial_tai_list[0]
+                            .u.tai_one_plmn_consecutive_tacs.plmn)) &&
+          (tai.tac == emm_ctx_tai->partial_tai_list[0]
+                          .u.tai_one_plmn_consecutive_tacs.tac)) {
```

Comment:
In case of "one plmn with consecutive tacs" there is only one tac present

Reply:
In case of "one plmn with consecutive tacs", there is only one plmn and one tac

Reply:
This is not an assumption. As per 24.301 9.9.3.33 if the type is "one plmn with consecutive tacs" i.e type of list -001 then only one tac value is filled. for e.g if the tac values 2,3,4 are configured, tac value will be 2 and the number of elements will be filled as 3. UE will interpret that there are totally 3 consecutive tacs starting from 2

Reply:
So if I take your example, let tai.tac = 3 or 4, the function should return true, but it does not ?

Reply:
This function still returns RETURNok because if the tai is not found in emm context then emm_verify_orig_tai() gets invoked which looks for the tai in mme config. MME config contains the complete TAI list which user has configured so it finds a match and returns RETURNok. But I guess I can use the first tac stored in emm context and the number of elements to check if the received tac falls in the range.

Reply:
> But I guess I can use the first tac stored in emm context and the number of elements to check if the received tac falls in the range.

Yes, so it will prevent MME from updating each time a TAU occurs the tai list in emm context, if UE does not move in a cell which is in the first tracking area